src/pixi_devenv/consolidate.py

```diff
@@ -0,0 +1,438 @@
+                return ":"
+            case unreachable:
+                assert_never(unreachable)
+
+
+def consolidate_devenv(workspace: Workspace) -> ConsolidatedProject:
+    """
+    Consolidates the given workspace definition, coalescing all pixi-devenv settings from the different files
+    in the workspace into a single pixi definition.
+    """
+    root_aspect = _consolidate_aspects(
+        workspace, [(p, p.get_root_aspect()) for p in workspace.iter_downstream()]
+    )
+
+    consolidated_target = _consolidate_target(workspace, list(workspace.iter_downstream()))
+    consolidated_feature = _consolidate_feature(workspace)
+
+    channels: tuple[str, ...] = ()
+    platforms: tuple[str, ...] = ()
+
+    for project in workspace.iter_downstream():
+        if project.channels:
+            channels = project.channels
+        if project.platforms:
+            platforms = project.platforms
+
+    return ConsolidatedProject(
+        name=workspace.starting_project.name,
+        channels=channels,
+        platforms=platforms,
+        dependencies=root_aspect.dependencies,
+        pypi_dependencies=root_aspect.pypi_dependencies,
+        env_vars=root_aspect.env_vars,
+        target=consolidated_target,
+        feature=consolidated_feature,
+    )
+
+
+@dataclass
+class ConsolidatedProject:
+    """
+    Result of consolidating all the projects in a `Workspace` in a single, final pixi.toml configuration.
+
+    This class acts as the schema that will be written as a `pixi.toml` file.
+    """
+
+    name: str
+
+    channels: tuple[str, ...]
+    platforms: tuple[str, ...]
+
+    dependencies: dict[str, MergedSpec]
+    pypi_dependencies: dict[str, MergedSpec]
+    env_vars: dict[str, MergedEnvVarValue]
+    target: dict[str, ConsolidatedAspect]
+    feature: dict[str, ConsolidatedFeature]
+
+
+type Sources = tuple[ProjectName, ...]
+
+
+@dataclass(frozen=True)
+class MergedSpec:
+    """
+    Specs from different projects merged together.
+
+    For example, we might have project "a" with:
+
+        pytest = ">=7.2"
+
+    And downstream, project "b" with:
+
+        pytest = { version = ">=8.0", channel = "conda-forge" }
+
+    This will result in a merged spec of:
+
+        pytest = { version = ">=7.2,>=8.0", build="", channel="conda-forge" }
+
+    which will be written in the final `pixi.toml`file.
+
+    Note:
+
+    * Versions are merged together.
+    * Build and channels cannot be merged, so the downstream-most project wins.
+    """
+
+    # Track which projects contributed to the spec.
+    sources: Sources
+
+    # Final merged spec.
+    spec: Spec
+
+    def add(self, spec_name: str, sources: ProjectName | tuple[ProjectName, ...], spec: Spec) -> MergedSpec:
+        if self.spec.version != "*" and spec.version != "*":
+            version = f"{self.spec.version},{spec.version}"
+        elif self.spec.version != "*":
+            version = self.spec.version
+        else:
+            version = spec.version
```

Comment:
What about the case that both declares the same version an that version is different of `*`. Wouldn't we end up with something like `==3.10,==3.10` eventually?

Reply:
Would `*,*` or `==3.10,*` be a problem? Because it kinda follows the same logic: if it wouldn't be troublesome, we wouldn't need this special handling. But if there's a gotcha, a code comment would be useful here.

Reply:
Ahh, OK, but I meant `*,*`, it's what would happen without the special treatment right? The code takes `*` from `self.spec.version`, `*` from `spec.version` and is outputting only `*`, without the treatment it would output `*,*`.

Reply:
Ahh OK, but yeah it would be fine as well: `*,*,==3.10` is equivalent to `==3.10`. 👍